packages/pwa-kit-dev/src/configs/webpack/overrides-plugin.js

```diff
@@ -0,0 +1,134 @@
+ *  allowing third party implementations that depend on an npm module for the base implementation
+ *  and then overriding only specific files
+ */
+class OverridesResolverPlugin {
+    /**
+     *
+     * @param options
+     * @param {string} options.overridesDir path to application base
+     * @param {string[]} options.overrides paths to overrides
+     * @param {string} options.projectDir path to project directory
+     */
+    constructor(options) {
+        this.overridesDir = options.overridesDir || ''
+        this.extends = options.extends || []
+        this.projectDir = options.projectDir
+        this._allSearchDirs = [this.projectDir + this.overridesDir, ...this.extends]
+        this.pkg = require(path.resolve(this.projectDir, 'package.json'))
+        this.extendsHashMap = new Map()
+
+        const OVERRIDES_EXTENSIONS = '.+(js|jsx|ts|tsx|svg|jpg|jpeg)'
+        const globPattern = `${this.pkg?.ccExtensibility?.overridesDir?.replace(
+            /\//,
+            ''
+        )}/**/*${OVERRIDES_EXTENSIONS}`
+        const overridesFsRead = glob.sync(globPattern)
+
+        const overrideReplace = this.pkg?.ccExtensibility?.overridesDir + '/'
+
+        overridesFsRead.forEach((item) => {
+            const end = item.substring(item.lastIndexOf('/index'))
+            const [l, ...rest] = item.split(/(index|\.)/)
+            this.extendsHashMap.set(
+                l.replace(/\/$/, '')?.replace(overrideReplace.replace(/\//, ''), ''),
+                [end, rest]
+            )
+        })
+    }
+
+    /**
+     *
+     * @param requestPath
+     * @param dirs
+     */
+    findFileMap(requestPath, dirs) {
+        var fileExt = path.extname(requestPath)
+        for (var dir of dirs) {
+            var base = path.join(dir, requestPath)
+            if (fileExt) {
+                const noExtPath = requestPath.replace(fileExt, '')
+                if (this.extendsHashMap.has(noExtPath)) {
+                    return base
+                }
+            } else {
+                if (this.extendsHashMap.has(requestPath)) {
+                    const end = this.extendsHashMap.get(requestPath)[1]
+
+                    if (end[0] === 'index') {
+                        base = path.join(base, this.extendsHashMap.get(requestPath)[1].join(''))
+                        return base
+                    } else {
+                        base = base?.replace(/$\//, '') + end.join('')
+                        return base
+                    }
+                }
+            }
+        }
+    }
+
+    toOverrideRelative(path) {
+        var override = this.findOverride(path)
+        return path.substring(override.length + 1)
+    }
+
+    findOverride(path) {
+        return this._allSearchDirs.find((override) => {
+            return path.indexOf(override) === 0
+        })
+    }
+
+    isFromExtends(request, path) {
+        return (
+            this.extends.includes(request?.split('/')?.[0]) &&
```

Comment:
(I don't know what is `request`, I assume it's the import statement string like `retail-react-app`) 

This doesn't work with name-spaced packages. for example `@salesforce/retail-react-app` 🤔 . 

For example, if request is `@salesforce/retail-react-app/app/ssr.js`, then `request?.split('/')?.[0]` is `@salesforce`...

Reply:
**BLOCKING** (I assume we're going to release v3 under @salesforce NPM org)

Reply:
Really great catch! :ok_hand: So... 🤔 

Here's my thinking: we can do a regex that says "if it begins with `@`, then treat the  **first** `/` character as part of the path"

I did some research and it looks like it's highly unlikely that `npm` will ever support nested package depth of greater than `1`. If that ever changes with `npm` it will be a breaking change for us.

But for now I'll assume we're safe. I did some research and the latest info indicates it's highly likely that we're safe

>  I would have said it was highly unlikely that the nesting of scopes would ever go beyond @<user or organization>/<package>, following the pattern established by GitHub, BitBucket, et al. But I couldn't really speak authoritatively about product priorities then and I definitely can't now.

https://github.com/npm/npm/issues/6333#issuecomment-300409124

Reply:
lgtm!